File: scripts/retriever.py
#!/usr/bin/env python3
"""Run PyLucene retrieval over FIRE topics and export a TSV run file.

Author: Shuvam Banerji Seal

Output format (tab-separated):
QID    Q0    DOCID    RANK    SCORE    RUNNAME

The retriever supports configurable similarity model + parameters.
Default model is BM25 with k1=1.5 and b=0.4 as required.
"""

from __future__ import annotations

import argparse
import builtins
import ssl
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable
from urllib.request import urlretrieve
from xml.etree import ElementTree as ET

import lucene
from java.nio.file import Paths
from org.apache.lucene.analysis.en import EnglishAnalyzer
from org.apache.lucene.analysis.standard import StandardAnalyzer
from org.apache.lucene.index import DirectoryReader
from org.apache.lucene.queryparser.classic import QueryParser
from org.apache.lucene.search import IndexSearcher
from org.apache.lucene.search.similarities import (
    BM25Similarity,
    ClassicSimilarity,
    LMDirichletSimilarity,
    LMJelinekMercerSimilarity,
)
from org.apache.lucene.store import MMapDirectory


PROJECT_ROOT = Path(__file__).resolve().parents[1]
DEFAULT_TOPICS_URL = "https://www.isical.ac.in/~fire/data/topics/adhoc/en.topics.176-225.2012.txt"
DEFAULT_TOPICS_FILE = "data/fire2012/adhoc/topics/en.topics.176-225.2012.txt"
DEFAULT_INDEX_DIR = "outputs/indexes/pylucene_en_docs"
DEFAULT_RUN_FILE = "outputs/runs/bm25_k1_1.5_b_0.4.tsv"

AUTHOR_NAME = "Shuvam Banerji Seal"


def log_print(*args, **kwargs) -> None:
    builtins.print(f"[Author: {AUTHOR_NAME}]", *args, **kwargs)


print = log_print


@dataclass
class Topic:
    qid: str
    title: str
    desc: str
    narr: str


def init_vm() -> None:
    vm_env = lucene.getVMEnv()
    if vm_env is None:
        lucene.initVM(vmargs=["-Djava.awt.headless=true"])
    else:
        vm_env.attachCurrentThread()


def resolve_path(p: str) -> Path:
    candidate = Path(p)
    if candidate.is_absolute():
        return candidate
    return PROJECT_ROOT / candidate


def normalize_text(value: str | None) -> str:
    if not value:
        return ""
    return " ".join(value.split())


def ensure_topics_file(path: Path, url: str, allow_insecure_ssl: bool) -> None:
    if path.exists():
        return

    path.parent.mkdir(parents=True, exist_ok=True)
    if allow_insecure_ssl:
        ssl._create_default_https_context = ssl._create_unverified_context

    print(f"[INFO] Downloading topics from: {url}")
    urlretrieve(url, str(path))
    print(f"[INFO] Topics saved to: {path}")


def parse_topics(topics_file: Path) -> list[Topic]:
    tree = ET.parse(topics_file)
    root = tree.getroot()

    topics: list[Topic] = []
    for top in root.findall(".//top"):
        qid = normalize_text(top.findtext("num"))
        title = normalize_text(top.findtext("title"))
        desc = normalize_text(top.findtext("desc"))
        narr = normalize_text(top.findtext("narr"))
        if not qid:
            continue
        topics.append(Topic(qid=qid, title=title, desc=desc, narr=narr))

    return topics


def build_similarity(model: str, k1: float, b: float, mu: float, lambd: float):
    if model == "bm25":
        return BM25Similarity(float(k1), float(b))
    if model == "classic":
        return ClassicSimilarity()
    if model == "lm_dirichlet":
        return LMDirichletSimilarity(float(mu))
    if model == "lm_jelinek_mercer":
        return LMJelinekMercerSimilarity(float(lambd))
    raise ValueError(f"Unsupported model: {model}")


def build_analyzer(name: str):
    if name == "english":
        return EnglishAnalyzer()
    if name == "standard":
        return StandardAnalyzer()
    raise ValueError(f"Unsupported analyzer: {name}")


def topic_to_query_text(topic: Topic, query_mode: str) -> str:
    if query_mode == "title":
        return topic.title
    if query_mode == "title_desc":
        return normalize_text(f"{topic.title} {topic.desc}")
    if query_mode == "title_desc_narr":
        return normalize_text(f"{topic.title} {topic.desc} {topic.narr}")
    raise ValueError(f"Unsupported query mode: {query_mode}")


def run_retrieval(
    *,
    topics: Iterable[Topic],
    index_dir: Path,
    output_file: Path,
    model: str,
    k1: float,
    b: float,
    mu: float,
    lambd: float,
    analyzer_name: str,
    default_operator: str,
    query_mode: str,
    top_k: int,
    run_name: str,
) -> None:
    init_vm()

    analyzer = build_analyzer(analyzer_name)
    similarity = build_similarity(model, k1, b, mu, lambd)

    directory = MMapDirectory(Paths.get(str(index_dir)))
    reader = DirectoryReader.open(directory)

    try:
        searcher = IndexSearcher(reader)
        searcher.setSimilarity(similarity)

        parser = QueryParser("content", analyzer)
        if default_operator == "AND":
            parser.setDefaultOperator(QueryParser.Operator.AND)
        else:
            parser.setDefaultOperator(QueryParser.Operator.OR)

        output_file.parent.mkdir(parents=True, exist_ok=True)

        query_count = 0
        written_rows = 0

        stored_fields = searcher.storedFields()

        with output_file.open("w", encoding="utf-8") as out:
            for topic in topics:
                query_count += 1
                query_text = topic_to_query_text(topic, query_mode)
                escaped_query = QueryParser.escape(query_text)
                if not escaped_query:
                    continue

                lucene_query = parser.parse(escaped_query)
                top_docs = searcher.search(lucene_query, top_k)

                for rank, score_doc in enumerate(top_docs.scoreDocs, start=1):
                    doc = stored_fields.document(score_doc.doc)
                    docid = doc.get("docno") or f"LUCENE_DOC_{score_doc.doc}"
                    out.write(
                        f"{topic.qid}\tQ0\t{docid}\t{rank}\t{score_doc.score:.6f}\t{run_name}\n"
                    )
                    written_rows += 1

        print("\n=== RETRIEVAL COMPLETE ===")
        print(f"index_dir: {index_dir}")
        print(f"topics_processed: {query_count}")
        print(f"rows_written: {written_rows}")
        print(f"run_file: {output_file}")
        print(f"model: {model}")
        if model == "bm25":
            print(f"bm25_k1: {k1}")
            print(f"bm25_b: {b}")
    finally:
        reader.close()
        directory.close()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Retrieve documents from PyLucene index")
    parser.add_argument("--index-dir", default=DEFAULT_INDEX_DIR, help="Path to Lucene index")
    parser.add_argument("--topics-file", default=DEFAULT_TOPICS_FILE, help="Path to FIRE topics XML")
    parser.add_argument("--topics-url", default=DEFAULT_TOPICS_URL, help="Topics download URL")
    parser.add_argument(
        "--download-topics-if-missing",
        action="store_true",
        help="Download topics file if --topics-file does not exist",
    )
    parser.add_argument(
        "--allow-insecure-ssl",
        action="store_true",
        help="Allow unverified SSL certificate while downloading topics",
    )
    parser.add_argument("--output-run", default=DEFAULT_RUN_FILE, help="Output TSV run file")
    parser.add_argument("--run-name", default="shuvam", help="Run name for the 6th column")
    parser.add_argument("--top-k", type=int, default=100, help="Number of docs to retrieve per topic")

    parser.add_argument(
        "--model",
        choices=["bm25", "classic", "lm_dirichlet", "lm_jelinek_mercer"],
        default="bm25",
        help="Retrieval similarity model",
    )
    parser.add_argument("--k1", type=float, default=1.5, help="BM25 k1")
    parser.add_argument("--b", type=float, default=0.4, help="BM25 b")
    parser.add_argument("--mu", type=float, default=2000.0, help="LM Dirichlet mu")
    parser.add_argument("--lambda", dest="lambd", type=float, default=0.2, help="LM JM lambda")

    parser.add_argument(
        "--analyzer",
        choices=["english", "standard"],
        default="english",
        help="Query analyzer",
    )
    parser.add_argument(
        "--default-operator",
        choices=["OR", "AND"],
        default="OR",
        help="Default boolean operator in QueryParser",
    )
    parser.add_argument(
        "--query-mode",
        choices=["title", "title_desc", "title_desc_narr"],
        default="title",
        help="How to build query text from topic fields",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()

    index_dir = resolve_path(args.index_dir)
    if not index_dir.exists():
        raise FileNotFoundError(f"Index directory not found: {index_dir}")

    topics_file = resolve_path(args.topics_file)
    if not topics_file.exists():
        if args.download_topics_if_missing:
            ensure_topics_file(topics_file, args.topics_url, args.allow_insecure_ssl)
        else:
            raise FileNotFoundError(
                f"Topics file not found: {topics_file}. "
                "Pass --download-topics-if-missing to fetch automatically."
            )

    topics = parse_topics(topics_file)
    if len(topics) != 50:
        print(f"[WARN] Expected 50 topics, found {len(topics)}")

    output_file = resolve_path(args.output_run)

    run_retrieval(
        topics=topics,
        index_dir=index_dir,
        output_file=output_file,
        model=args.model,
        k1=args.k1,
        b=args.b,
        mu=args.mu,
        lambd=args.lambd,
        analyzer_name=args.analyzer,
        default_operator=args.default_operator,
        query_mode=args.query_mode,
        top_k=args.top_k,
        run_name=args.run_name,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: java_lucene_implementation/src/main/java/cs4201/Retriever.java
package cs4201;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.en.EnglishAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.index.StoredFields;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.TopDocs;
import org.apache.lucene.search.similarities.BM25Similarity;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.MMapDirectory;

import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import javax.xml.parsers.DocumentBuilderFactory;

import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * CS4201 Information Retrieval — Java Lucene Retriever
 * Author : Shuvam Banerji Seal (22MS076)
 * Course Instructor : Dr. Dwaipayan Roy
 *
 * Queries a Lucene index built by Indexer.java over FIRE 2012 English Ad-hoc topics
 * and writes a TREC-style 6-column TSV run file.
 *
 * Configuration:
 *   Similarity  : BM25(k1=1.5, b=0.4)
 *   Analyzer    : EnglishAnalyzer (Porter stemming + stopwords)
 *   Query mode  : title-only
 *   Operator    : OR
 *   Top-K       : 100 documents per topic
 *
 * Output format (tab-separated, 6 columns):
 *   QID   Q0   DOCID   RANK   SCORE   RUNNAME
 *
 * Usage (from java_lucene_implementation/ directory):
 *   java -cp target/java-lucene-ir-1.0-SNAPSHOT.jar cs4201.Retriever \
 *        [--index-dir <path>] [--topics-file <path>] [--output-run <path>] \
 *        [--run-name <name>] [--top-k <n>]
 */
public class Retriever {

    // BM25 parameters — k1 controls term-frequency saturation, b controls length normalisation.
    static final float BM25_K1 = 1.5f;
    static final float BM25_B  = 0.4f;

    // Default paths (relative to CWD = java_lucene_implementation/)
    static final String DEFAULT_INDEX_DIR  = "../outputs/indexes/java_lucene_en_docs";
    static final String DEFAULT_TOPICS     = "../data/fire2012/adhoc/topics/en.topics.176-225.2012.txt";
    static final String DEFAULT_OUTPUT     = "../outputs/runs/java_bm25_k1_1.5_b_0.4_Shuvam_Banerji_Seal_22MS076.tsv";
    static final String DEFAULT_RUN_NAME   = "shuvam_java_bm25";
    static final int    DEFAULT_TOP_K      = 100;

    // Topic record — holds the four FIRE topic fields.
    record Topic(String qid, String title, String desc, String narr) {}

    public static void main(String[] args) throws Exception {
        String indexDir   = DEFAULT_INDEX_DIR;
        String topicsFile = DEFAULT_TOPICS;
        String outputFile = DEFAULT_OUTPUT;
        String runName    = DEFAULT_RUN_NAME;
        int    topK       = DEFAULT_TOP_K;

        // Parse command-line arguments — manual switch, no external library needed.
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--index-dir"   -> indexDir   = args[++i];
                case "--topics-file" -> topicsFile = args[++i];
                case "--output-run"  -> outputFile = args[++i];
                case "--run-name"    -> runName    = args[++i];
                case "--top-k"       -> topK       = Integer.parseInt(args[++i]);
                default -> System.err.println("[WARN] Unknown argument: " + args[i]);
            }
        }

        Path indexPath  = Indexer.resolveRelativePath(indexDir);
        Path topicsPath = Indexer.resolveRelativePath(topicsFile);
        Path outputPath = Indexer.resolveRelativePath(outputFile);

        // Sanity checks — fail fast with a useful message rather than a cryptic NPE.
        if (!Files.exists(indexPath)) {
            System.err.println("[ERROR] Index directory not found: " + indexPath);
            System.err.println("        Please run Indexer first:");
            System.err.println("        java -cp target/*.jar cs4201.Indexer");
            System.exit(1);
        }
        if (!Files.exists(topicsPath)) {
            System.err.println("[ERROR] Topics file not found: " + topicsPath);
            System.exit(1);
        }

        System.out.println("[CS4201-Java-Retriever] BM25 k1=" + BM25_K1 + "  b=" + BM25_B);
        System.out.println("[CS4201-Java-Retriever] Index:   " + indexPath);
        System.out.println("[CS4201-Java-Retriever] Topics:  " + topicsPath);
        System.out.println("[CS4201-Java-Retriever] Output:  " + outputPath);
        System.out.println("[CS4201-Java-Retriever] Run name: " + runName);

        // Parse FIRE 2012 topics from XML.
        List<Topic> topics = parseTopics(topicsPath);
        System.out.println("[CS4201-Java-Retriever] Loaded " + topics.size() + " topics");
        if (topics.size() != 50) {
            System.err.println("[WARN] Expected 50 topics, got " + topics.size());
        }

        // Open the Lucene index (read-only).
        Directory directory = MMapDirectory.open(indexPath);
        DirectoryReader reader = DirectoryReader.open(directory);

        System.out.println("[CS4201-Java-Retriever] Index has " + reader.numDocs() + " documents");

        try {
            IndexSearcher searcher = new IndexSearcher(reader);
            // Similarity must match what was used at index time.
            searcher.setSimilarity(new BM25Similarity(BM25_K1, BM25_B));

            // EnglishAnalyzer — same stemmer and stopwords used at index time.
            Analyzer analyzer = new EnglishAnalyzer();

            // Query targets the "content" field (title + body concatenated at index time).
            QueryParser parser = new QueryParser("content", analyzer);

            // OR operator maximises recall; empirically best for FIRE 2012 short titles.
            parser.setDefaultOperator(QueryParser.Operator.OR);

            // Lucene 9.5+ stored-fields API.
            StoredFields storedFields = searcher.storedFields();

            // Ensure output directory exists before trying to write there.
            Files.createDirectories(outputPath.getParent());

            int totalWritten = 0;

            try (PrintWriter out = new PrintWriter(new BufferedWriter(
                    new OutputStreamWriter(
                            new FileOutputStream(outputPath.toFile()),
                            StandardCharsets.UTF_8)))) {

                for (Topic topic : topics) {
                    // Title-only queries; description+narrative hurt MAP on these topics.
                    String queryText = Indexer.normalizeText(topic.title());
                    if (queryText.isEmpty()) {
                        System.err.println("[WARN] Empty title for topic " + topic.qid() + ", skipping");
                        continue;
                    }

                    // Escape special Lucene syntax chars (e.g. +, :, ~) to prevent parse errors.
                    String escapedQuery = QueryParser.escape(queryText);
                    if (escapedQuery.trim().isEmpty()) continue;

                    // Parse the escaped query string into a Lucene Query object.
                    Query query;
                    try {
                        query = parser.parse(escapedQuery);
                    } catch (Exception e) {
                        System.err.println("[WARN] Query parse failed for topic " + topic.qid()
                                + " (\"" + queryText + "\"): " + e.getMessage());
                        continue;
                    }

                    TopDocs topDocs = searcher.search(query, topK);
                    ScoreDoc[] hits  = topDocs.scoreDocs;

                    // Write one TSV row per hit in TREC run format.
                    for (int rank = 1; rank <= hits.length; rank++) {
                        ScoreDoc hit = hits[rank - 1];

                        Document doc = storedFields.document(hit.doc);
                        String docId = doc.get("docno");

                        if (docId == null || docId.isEmpty()) {
                            docId = "LUCENE_DOC_" + hit.doc;
                        }

                        out.print(topic.qid() + "\tQ0\t" + docId + "\t" + rank
                                + "\t" + String.format("%.6f", hit.score)
                                + "\t" + runName + "\n");
                        totalWritten++;
                    }
                }

            }

            System.out.println("\n=== RETRIEVAL COMPLETE ===");
            System.out.println("topics_processed: " + topics.size());
            System.out.println("rows_written:     " + totalWritten);
            System.out.println("run_file:         " + outputPath);
            System.out.printf( "model:            BM25(k1=%.1f, b=%.1f)%n", BM25_K1, BM25_B);
            System.out.println("run_name:         " + runName);

        } finally {
            reader.close();
            directory.close();
        }
    }

    /**
     * Parses the FIRE 2012 topics XML file and returns a list of Topic records.
     * Uses the JDK DOM parser with XXE protections enabled.
     */
    static List<Topic> parseTopics(Path topicsFile) throws Exception {
        // Disable external entity resolution to prevent XXE injection.
        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", false);
        dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
        dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
        dbf.setExpandEntityReferences(false);

        var db  = dbf.newDocumentBuilder();
        var doc = db.parse(topicsFile.toFile());

        List<Topic> topics = new ArrayList<>();
        NodeList topNodes = doc.getElementsByTagName("top");

        for (int i = 0; i < topNodes.getLength(); i++) {
            Element top = (Element) topNodes.item(i);

            String qid   = Indexer.normalizeText(tagText(top, "num"));
            String title = Indexer.normalizeText(tagText(top, "title"));
            String desc  = Indexer.normalizeText(tagText(top, "desc"));
            String narr  = Indexer.normalizeText(tagText(top, "narr"));

            if (qid.isEmpty()) {
                System.err.println("[WARN] Topic at index " + i + " has no <num> element, skipping");
                continue;
            }

            topics.add(new Topic(qid, title, desc, narr));
        }

        return topics;
    }

    /**
     * Return the text content of the first child element with the given tag name,
     * or "" if not found.  Equivalent to Python's {@code top.findtext(tagName) or ""}.
     */
    private static String tagText(Element parent, String tagName) {
        NodeList nodes = parent.getElementsByTagName(tagName);
        if (nodes.getLength() == 0) return "";
        return nodes.item(0).getTextContent();
    }
}
